data/finset.lean

```diff
@@ -899,6 +904,156 @@ sort_eq _ _
 
 @[simp] theorem sort_to_finset [decidable_eq α] (s : finset α) : (sort r s).to_finset = s :=
 list.to_finset_eq (sort_nodup r s) ▸ eq_of_veq (sort_eq r s)
+
 end sort
 
+def disjoint (s t : finset α) : Prop := ∀ ⦃a⦄, a ∈ s → a ∈ t → false
+
+theorem disjoint.symm {s t : finset α} (d : disjoint s t) : disjoint t s
+| a i₂ i₁ := d i₁ i₂
+
+@[simp] theorem disjoint_comm {s t : finset α} : disjoint s t ↔ disjoint t s :=
+⟨disjoint.symm, disjoint.symm⟩
+
+theorem disjoint_left {s t : finset α} : disjoint s t ↔ ∀ {a}, a ∈ s → a ∉ t := iff.rfl
+
+theorem disjoint_right {s t : finset α} : disjoint s t ↔ ∀ {a}, a ∈ t → a ∉ s :=
+disjoint_comm
+
+theorem disjoint_iff_ne {s t : finset α} : disjoint s t ↔ ∀ a ∈ s, ∀ b ∈ t, a ≠ b :=
+by simp [disjoint_left, imp_not_comm]
+
+theorem disjoint_of_subset_left {s t u : finset α} (h : s ⊆ u) (d : disjoint u t) : disjoint s t
+| x m₁ := d (h m₁)
+
+theorem disjoint_of_subset_right {s t u : finset α} (h : t ⊆ u) (d : disjoint s u) : disjoint s t
+| x m m₁ := d m (h m₁)
+
+@[simp] theorem empty_disjoint (l : finset α) : disjoint ∅ l
+| a := (not_mem_empty a).elim
+
+@[simp] theorem singleton_disjoint [decidable_eq α] {l : finset α} {a : α} : disjoint (singleton a) l ↔ a ∉ l :=
+by simp [disjoint]; refl
+
+@[simp] theorem disjoint_singleton [decidable_eq α] {l : finset α} {a : α} : disjoint l (singleton a) ↔ a ∉ l :=
+by rw disjoint_comm; simp
+
+@[simp] theorem disjoint_insert_left [decidable_eq α] {a : α} {s t : finset α} :
+  disjoint (insert a s) t ↔ a ∉ t ∧ disjoint s t :=
+by simp [disjoint, or_imp_distrib, forall_and_distrib]; refl
+
+@[simp] theorem disjoint_insert_right [decidable_eq α] {a : α} {s t : finset α} :
+  disjoint s (insert a t) ↔ a ∉ s ∧ disjoint s t :=
+disjoint_comm.trans $ by simp [disjoint_insert_left]
+
+theorem inter_eq_empty_iff_disjoint [decidable_eq α] {s t : finset α} : s ∩ t = ∅ ↔ disjoint s t :=
+by simp [ext, mem_inter]; refl
+
+theorem inter_eq_zero_iff_disjoint [decidable_eq α] {s t : finset α} : s ∩ t = ∅ ↔ disjoint s t :=
+by rw ← subset_empty; simp [subset_iff, disjoint]
+
+@[simp] theorem disjoint_union_left [decidable_eq α] {s t u : finset α} :
+  disjoint (s ∪ t) u ↔ disjoint s u ∧ disjoint t u :=
+by simp [disjoint, or_imp_distrib, forall_and_distrib]
+
+@[simp] theorem disjoint_union_right [decidable_eq α] {s t u : finset α} :
+  disjoint s (t ∪ u) ↔ disjoint s t ∧ disjoint s u :=
+by simp [disjoint, or_imp_distrib, forall_and_distrib]
+
+@[simp] theorem disjoint_val {s t : finset α} : multiset.disjoint s.1 t.1 ↔ disjoint s t :=
+by simp [disjoint, multiset.disjoint, mem_def]
+
+@[simp] theorem card_disjoint_union [decidable_eq α] {s t : finset α} :
+    disjoint s t → card (s ∪ t) = card s + card t :=
+finset.induction_on s (by simp) $ begin simp {contextual := tt} end
+
+end finset
+
+namespace list
+open finset
+variable [decidable_eq α]
+ 
+definition powerset : list α → finset (finset α)
+| []       := {∅}
+| (a :: l) := powerset l ∪ image (insert a) (powerset l)
```

Comment:
This function already exists, it is called `sublists`. I would prefer to see the proof that `sublists` is in fact the list function you need to construct the finset function `powerset`.

Reply:
How about just using `multiset.rec_on` to define it for multisets, and then defining it for finsets in terms of that? I did try lifting sublists, but it's a lot messier than `multiset.rec_on`.

Thanks for proving all of the stuff about sublists.
Out of interest, is there any reason you didn't want to use this definition for sublists.
```lean
def sublists' : list α → list (list α) 
| list.nil := [list.nil]
| (a :: l) := list.map (list.cons a) (sublists' l) ++ sublists' l
```
It's a lot easier to work with, as I can use perm.induction. I'm guessing there is a good reason, but it does seem a bit odd.

Reply:
I was thinking about defining that as well, for analysis purposes. The reason is efficiency: the repeated maps and appends there are much slower than doing it by tail recursion. (Remember that `list` is also a programming library, and `finset` is still in that computable fragment.)